src/plutus_gen/code_emitters_aiken.rs

```diff
@@ -901,44 +916,142 @@ impl ElementMSM {
+                ElementMSM::ElementNegatedG1(scalar) => (G1PointKey::NegatedG1, scalar),
+            };
+
+            // Track insertion order for deterministic output
+            if !groups.contains_key(&key) {
+                insertion_order.push(key.clone());
+            }
+
+            groups.entry(key).or_insert_with(Vec::new).push(scalar);
+        }
+
+        // Combine scalars for each G1 point
+        let optimized_elements: Vec<ElementMSM> = insertion_order
+            .into_iter()
+            .map(|key| {
+                let scalars = groups.remove(&key).unwrap();
+
+                // Combine all scalars by adding them together
+                let combined_scalar = scalars.into_iter().reduce(|acc, scalar| {
+                    ScalarOperation::Add(Box::new(acc), Box::new(scalar))
+                }).unwrap();
+
+                // Reconstruct the element with combined scalar
+                match key {
+                    G1PointKey::Commitment(commitment) => {
+                        ElementMSM::Element(combined_scalar, commitment)
+                    }
+                    G1PointKey::W(index) => {
+                        ElementMSM::ElementW(combined_scalar, index)
+                    }
+                    G1PointKey::NegatedG1 => {
+                        ElementMSM::ElementNegatedG1(combined_scalar)
+                    }
+                }
+            })
+            .collect();
+
+        OptimizedMSM {
+            elements: optimized_elements,
+        }
+    }
+
+    /// Finds the maximum power exponent for a given variable in an MSM.
+    /// Recursively traverses all scalar operations to find Power(var_name, exponent).
+    fn find_max_power(&self, var_name: char) -> i32 {
+        (*self).elements
+            .iter()
+            .map(|element| {
+                let scalar = match element {
+                    ElementMSM::Element(s, _) => s,
+                    ElementMSM::ElementW(s, _) => s,
+                    ElementMSM::ElementNegatedG1(s) => s,
+                };
+                Self::find_max_power_in_scalar(scalar, var_name)
+            })
+            .max()
+            .unwrap_or(0)
+    }
+
+    /// Recursively finds max power exponent in a scalar operation tree
+    fn find_max_power_in_scalar(scalar: &ScalarOperation, var_name: char) -> i32 {
+        match scalar {
+            ScalarOperation::Power(name, exponent) if *name == var_name => *exponent,
+            ScalarOperation::Mul(s, _) => Self::find_max_power_in_scalar(s, var_name),
+            ScalarOperation::MulS(s1, s2) => {
+                Self::find_max_power_in_scalar(s1, var_name).max(Self::find_max_power_in_scalar(s2, var_name))
+            }
+            ScalarOperation::Add(s1, s2) => {
+                Self::find_max_power_in_scalar(s1, var_name).max(Self::find_max_power_in_scalar(s2, var_name))
+            }
+            _ => 0,
```

Comment:
@adamsmo I refactored find_max_power to filter MSMElements instead of using regex. Although I'm not sure it's better. Now more compex logic is involved, because I need to traverse the Scalar tree. 
Do you think this solution is better?

Reply:
from my perspective it is better the regexp on final transpiled code because it is type safe and this eliminates entire class of errors
it is traversing a tree, which is easier to follow the a regex on blob of text, when you read the code
and true, the code is more complex, but the complexity is at code generation step, it does not affect cardano runtime